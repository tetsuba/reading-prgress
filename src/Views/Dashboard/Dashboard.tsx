import { useSelector } from 'react-redux'
import { useQuery } from 'react-query'
import { userIdSelector } from '../../store/user/userSelectors'
import { getWords } from '../../lib/service'

// COMPONENTS
import Header from '../../Components/Header/Header'
import HeatMap from '../../Components/HeatMap/HeatMap'
import Banner from '../../Components/Banner/Banner'
import Loading from '../../Components/Loading/Loading'

type LastBookReadTypes = {
    words: string[]
    date: string
    title: string
}

export default function Dashboard() {
    const userId = useSelector(userIdSelector)
    const { data, isSuccess, isLoading } = useQuery(['words', userId], getWords)

    const getColor = (book: LastBookReadTypes) =>
        book.words.length ? 'red' : 'green'

    if (isLoading) {
        return <Loading />
    }
    return (
        <>
            <Header text="Dashboard" />
            <main>
                <div className="mx-auto max-w-7xl py-6 sm:px-6 lg:px-8">
                    <div className="px-4 sm:px-0">
                        {isSuccess && (
                            <>
                                <div>
                                    {data.data.lastBookRead.map(
                                        (
                                            book: LastBookReadTypes,
                                            i: number
                                        ) => (
                                            <Banner
                                                key={`banner-${i}`}
                                                className="mt-4"
                                                color={getColor(book)}
                                            >
                                                [{book.date}] {book.title}{' '}
                                            </Banner>
                                        )
                                    )}
                                </div>
                                <HeatMap
                                    color="red"
                                    words={data.data.readIncorrectly.oneWeekAgo}
                                    search={''}
                                >
                                    Words read incorrectly (In the last week)
                                </HeatMap>
                                <HeatMap
                                    color="red"
                                    words={
                                        data.data.readIncorrectly.oneMonthAgo
                                    }
                                    search={''}
                                >
                                    Words read incorrectly (In the last month)
                                </HeatMap>
                                <HeatMap
                                    color="red"
                                    words={data.data.readIncorrectly.history}
                                    search={''}
                                >
                                    Words read incorrectly (More than a month
                                    ago)
                                </HeatMap>
                            </>
                        )}
                    </div>
                </div>
            </main>
        </>
    )
}
