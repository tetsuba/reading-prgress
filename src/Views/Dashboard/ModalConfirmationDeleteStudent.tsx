import { useMutation } from 'react-query'
import * as R from 'ramda'

// API
import { deleteStudent } from '../../api/api-students'

// COMPONENTS
import Modal from '../../Components/Modal/Modal'
import Confirmation from '../../Components/Modal/Confirmation'
import Display from '../../Components/Dispay/Display'
import { useDispatch } from 'react-redux'
import { addStudents } from '../../store/students/studentsSlice'
import { StateStudentTypes } from '../../store/store-types'

// TODO: if delete responds an error what should happen?

type PropTypes = {
    student: StateStudentTypes | null
    closeModal: () => void
    show: boolean
}

export default function ModalConfirmationDeleteStudent(props: PropTypes) {
    const dispatch = useDispatch()
    const { student, show, closeModal } = props
    const mutation = useMutation(deleteStudent, {
        onSuccess: (data) => {
            dispatch(addStudents(data.data))
            closeModal()
        }
    })

    return (
        <Display value={show}>
            <Modal className="max-w-md">
                <Confirmation
                    bookTitle={`  
                        ${R.pathOr('', ['firstname'], student)} 
                        ${R.pathOr('', ['lastname'], student)} 
                    `}
                    clickHandlerCancel={() => closeModal()}
                    clickHandlerDelete={() => {
                        if (student) mutation.mutate(student.studentId)
                    }}
                />
            </Modal>
        </Display>
    )
}
