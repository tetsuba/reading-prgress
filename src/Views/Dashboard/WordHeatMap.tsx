export type WordHeatMapTypes = {
    word: string
    index: number
}

function getHeatMap(index: number) {
    if (index < 2) {
        return { number: '', bg: 'border-gray-100', text: 'text-gray-900' }
    } else if (index < 5) {
        return {
            number: 'bg-red-100',
            bg: 'border-gray-100 bg-red-50',
            text: 'text-gray-900'
        }
    } else if (index < 10) {
        return {
            number: 'bg-red-300',
            bg: 'border-gray-100 bg-red-100',
            text: 'text-gray-800'
        }
    } else if (index < 15) {
        return {
            number: 'bg-red-400',
            bg: 'border-gray-200 bg-red-100',
            text: 'text-gray-700'
        }
    } else if (index < 20) {
        return {
            number: 'bg-red-500',
            bg: 'border-gray-300 bg-red-200',
            text: 'text-gray-600'
        }
    } else {
        return {
            number: 'bg-black',
            bg: 'border-gray-300 bg-red-500',
            text: 'text-white'
        }
    }
}

export default function WordHeatMap(props: WordHeatMapTypes) {
    const heatMap = getHeatMap(props.index)

    return (
        <span
            data-testid="heat-map-word"
            className={`m-6 rounded-2xl border-2 p-3 ${heatMap.bg}`}
        >
            <span className={`font-bold text-gray-900 ${heatMap.text}`}>
                {props.word}
            </span>
            <span
                className={`ml-4 rounded-full py-1 px-3 text-white ${heatMap.number}`}
            >
                {props.index}
            </span>
        </span>
    )
}
