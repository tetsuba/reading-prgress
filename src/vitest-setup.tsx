import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
