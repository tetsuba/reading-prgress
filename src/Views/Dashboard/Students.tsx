import { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import * as R from 'ramda'

// STORE
import { studentsSelector } from '../../store/students/studentsSelectors'
import { updateViewStudentId } from '../../store/view/viewSlice'

// TYPES
import { StateStudentTypes } from '../../store/store-types'

// COMPONENTS
import Loop from '../../Components/Loop/Loop'
import StudentRow from '../../Components/Row/StudentRow'
import ModalConfirmationDeleteStudent from './ModalConfirmationDeleteStudent'

export default function Students() {
    const dispatch = useDispatch()
    const students = useSelector(studentsSelector)
    const [deleteStudent, setDeleteStudent] =
        useState<StateStudentTypes | null>(null)

    return (
        <section>
            <Loop array={students}>
                <StudentRow
                    selectStudent={(student) => {
                        dispatch(updateViewStudentId(student.studentId))
                    }}
                    deleteStudent={(student) => {
                        setDeleteStudent(student)
                    }}
                />
            </Loop>
            <ModalConfirmationDeleteStudent
                student={deleteStudent}
                closeModal={() => setDeleteStudent(null)}
                show={R.isNotNil(deleteStudent)}
            />
        </section>
    )
}
