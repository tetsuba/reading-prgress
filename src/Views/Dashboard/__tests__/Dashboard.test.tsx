import { render, screen, waitFor } from '@testing-library/react'
import { Mocked } from 'vitest'
import { WrapperWith_Store_Query_Router } from '../../../vitest-setup'
import Dashboard from '../Dashboard'

import axios from 'axios'
vi.mock('axios')
const mockedAxios = axios as Mocked<typeof axios>

const mockData = {
    data: {
        lastBookRead: [
            { date: '12/12/2023', title: 'The Perfect Pet', words: [] },
            { date: '12/12/2023', title: 'Double Trouble', words: ['there'] }
        ],
        readIncorrectly: {
            oneWeekAgo: [
                { word: 'this', index: 1 },
                { word: 'that', index: 5 }
            ],
            oneMonthAgo: [
                { word: 'would', index: 10 },
                { word: 'dog', index: 15 }
            ],
            history: [
                { word: 'right', index: 20 },
                { word: 'live', index: 40 }
            ]
        }
    }
}

describe('Dashboard', () => {
    test('should render a list of words', async () => {
        mockedAxios.get.mockResolvedValue(mockData)
        const { asFragment } = render(
            <WrapperWith_Store_Query_Router pathname={'/dashboard'}>
                <Dashboard />
            </WrapperWith_Store_Query_Router>
        )
        await waitFor(() =>
            expect(screen.queryAllByTestId('heat-map-word')).toHaveLength(6)
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
