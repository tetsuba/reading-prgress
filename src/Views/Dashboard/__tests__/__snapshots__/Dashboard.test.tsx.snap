// Vitest Snapshot v1

exports[`Dashboard > should render a list of words 1`] = `
<DocumentFragment>
  <header
    class="bg-white shadow"
  >
    <div
      class="mx-auto max-w-7xl py-6 px-4 sm:px-6 lg:px-8"
    >
      <h1
        class="text-3xl font-bold tracking-tight text-gray-900"
      >
        Dashboard
      </h1>
    </div>
  </header>
  <main>
    <div
      class="mx-auto max-w-7xl py-6 sm:px-6 lg:px-8"
    >
      <div
        class="px-4 py-6 sm:px-0"
      >
        <h3
          class="text-xl font-medium text-gray-900 dark:text-white mt-10 rounded-t-lg bg-gray-200 p-2"
        >
          Words read incorrectly
        </h3>
        <div
          class="min-h-96 inline-flex w-full flex-wrap items-center justify-center rounded-b-lg border-4 border-t-0 border-dashed border-gray-200 p-2 shadow-md"
          data-testid="heat-map"
        >
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-100"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-gray-900"
            >
              this
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white "
            >
              1
            </span>
          </span>
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-100 bg-red-100"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-gray-800"
            >
              that
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white bg-red-300"
            >
              5
            </span>
          </span>
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-200 bg-red-100"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-gray-700"
            >
              would
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white bg-red-400"
            >
              10
            </span>
          </span>
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-300 bg-red-200"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-gray-600"
            >
              dog
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white bg-red-500"
            >
              15
            </span>
          </span>
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-300 bg-red-500"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-white"
            >
              right
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white bg-black"
            >
              20
            </span>
          </span>
          <span
            class="m-6 rounded-2xl border-2 p-3 border-gray-300 bg-red-500"
            data-testid="heat-map-word"
          >
            <span
              class="font-bold text-gray-900 text-white"
            >
              live
            </span>
            <span
              class="ml-4 rounded-full py-1 px-3 text-white bg-black"
            >
              40
            </span>
          </span>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
